File: src/ios/epos2Plugin.m
#import "ePOS2Plugin.h"

#import <Cordova/CDVAvailability.h>

@interface epos2Plugin()<Epos2DiscoveryDelegate, Epos2PtrReceiveDelegate>
@end

@implementation epos2Plugin

- (void)pluginInitialize {
    printerSeries = EPOS2_TM_U220;
    lang = EPOS2_MODEL_ANK;
}

- (void)startDiscover:(CDVInvokedUrlCommand *)command {
    self.discoverCallbackId = command.callbackId;

    Epos2FilterOption *filteroption_ = [[Epos2FilterOption alloc] init];
    [filteroption_ setDeviceType:EPOS2_TYPE_PRINTER];

    int result = [Epos2Discovery start:filteroption_ delegate:self];
    if (EPOS2_SUCCESS != result) {
        NSLog(@"error in startDiscover()");
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in discovering printer."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
    }
}

- (void) onDiscovery:(Epos2DeviceInfo *)deviceInfo {
    NSLog(@"onDiscovery: %@", [deviceInfo getIpAddress]);
    NSDictionary *info = @{@"ipAddress" : [deviceInfo getIpAddress],
                           @"deviceType": [NSNumber numberWithInt:[deviceInfo getDeviceType]],
                           @"target": [deviceInfo getTarget],
                           @"deviceName": [deviceInfo getDeviceName],
                           @"macAddress": [deviceInfo getMacAddress],
                           @"bdAddress": [deviceInfo getBdAddress],
                           };
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:info];
    [result setKeepCallbackAsBool:YES];
    [self.commandDelegate sendPluginResult:result callbackId:self.discoverCallbackId];
}

- (void)stopDiscover:(CDVInvokedUrlCommand *)command {
    NSLog(@"stopDiscover()");
    int result = EPOS2_SUCCESS;

    while (YES) {
        result = [Epos2Discovery stop];

        if (result != EPOS2_ERR_PROCESSING) {
            break;
        }
    }

    if (EPOS2_SUCCESS != result) {
        NSLog(@"error in stopDiscover()");
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in stop discovering printer."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
    }
}

-(void)connectPrinter:(CDVInvokedUrlCommand *)command {
    NSString* target = [command.arguments objectAtIndex:0];
    NSLog(@"connectPrinter(), ip=%@", target);

    printer = [[Epos2Printer alloc] initWithPrinterSeries:printerSeries lang:lang];

    [printer setReceiveEventDelegate:self];

    int result = EPOS2_SUCCESS;

    if (printer == nil) {
        NSLog(@"error in connectPrinter(), printer object not found");
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in connect  printer. Printer object not found"];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        return;
    }

    result = [printer connect:[NSString stringWithFormat:@"TCP:%@", target] timeout:EPOS2_PARAM_DEFAULT];
    if (result != EPOS2_SUCCESS) {
        NSLog(@"error in connectPrinter()");
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in connect  printer."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        return;
    }

    result = [printer beginTransaction];
    if (result != EPOS2_SUCCESS) {
        NSLog(@"error in beginTransaction()");
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in begin transaction."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        [printer disconnect];
        return;
    }

    CDVPluginResult *cordovaResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsBool:YES];
    [self.commandDelegate sendPluginResult:cordovaResult callbackId:command.callbackId];
}

- (void)disconnectPrinter:(CDVInvokedUrlCommand *)command {
    int result = EPOS2_SUCCESS;
    //    NSMutableDictionary *dict = [[NSMutableDictionary alloc] init];

    if (printer == nil) {
        return;
    }

    result = [printer endTransaction];
    if (result != EPOS2_SUCCESS) {
        NSLog(@"error in disconnectPrinter()");
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in end transaction."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        //        [dict setObject:[NSNumber numberWithInt:result] forKey:@"result"];
        //        [dict setObject:@"endTransaction" forKey:@"method"];
        //        [self performSelectorOnMainThread:@selector(showEposErrorFromThread:) withObject:dict waitUntilDone:NO];
    }

    result = [printer disconnect];
    if (result != EPOS2_SUCCESS) {
        NSLog(@"error in disconnectPrinter()");
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in disconnect printer."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        //        [dict setObject:[NSNumber numberWithInt:result] forKey:@"result"];
        //        [dict setObject:@"disconnect" forKey:@"method"];
        //        [self performSelectorOnMainThread:@selector(showEposErrorFromThread:) withObject:dict waitUntilDone:NO];
    }
    [self finalizeObject];

    CDVPluginResult *cordovaResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsBool:YES];
    [self.commandDelegate sendPluginResult:cordovaResult callbackId:command.callbackId];
}

- (void)finalizeObject
{
    if (printer == nil) {
        return;
    }

    [printer clearCommandBuffer];

    [printer setReceiveEventDelegate:nil];

    printer = nil;
}

- (void)onPtrReceive:(Epos2Printer *)printerObj code:(int)code status:(Epos2PrinterStatusInfo *)status printJobId:(NSString *)printJobId {
    NSLog(@"onPtrReceive");
    NSLog(@"code: %d", code);
    NSLog(@"status: %@", status);
    NSLog(@"printJobId: %@", printJobId);

    [self disconnectPrinter:nil];
}

- (void)print:(CDVInvokedUrlCommand *)command {
    self.printCallbackId = command.callbackId;

    if (printer == nil) {
        NSLog(@"error in createPrintData()");
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in create data, printer object not dound."];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        return;
    }

    NSArray *printData = command.arguments;

    [self.commandDelegate runInBackground:^{
        int result = EPOS2_SUCCESS;

        for (NSString* data in printData) {
            NSLog(@"%@", data);
            if ([data isEqualToString:@"\n"]) {
                result = [printer addFeedLine:1];
            } else {
                result = [printer addText:data];
            }

            if (result != EPOS2_SUCCESS) {
                break;
            }
        }

        if (result != EPOS2_SUCCESS) {
            return;
        }

        [self printData];
    }];
}

- (void)printData {
    int result = EPOS2_SUCCESS;

//    Epos2PrinterStatusInfo *status = nil;

//    if (printer == nil) {
//        return;
//    }
//
//    if (![self connectPrinter]) {
//        return;
//    }

//    status = [printer getStatus];
//    [self dispPrinterWarnings:status];

//    if (![self isPrintable:status]) {
//        [ShowMsg show:[self makeErrorMessage:status]];
//        [printer disconnect];
//        return;
//    }

    result = [printer sendData:EPOS2_PARAM_DEFAULT];
    if (result != EPOS2_SUCCESS) {
        [printer disconnect];

        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Error in print data."];
        [self.commandDelegate sendPluginResult:result callbackId:self.printCallbackId];
        return;
    }

    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsBool:YES];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:self.printCallbackId];
    return;
}

@end

File: src/android/epos2Plugin.java
/**
 */
package com.cordova.plugin;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CordovaWebView;
import org.apache.cordova.PluginResult;
import org.apache.cordova.PluginResult.Status;
import org.json.JSONObject;
import org.json.JSONArray;
import org.json.JSONException;

import android.util.Log;

import com.epson.epos2.Epos2Exception;
import com.epson.epos2.discovery.DeviceInfo;
import com.epson.epos2.discovery.Discovery;
import com.epson.epos2.discovery.DiscoveryListener;
import com.epson.epos2.discovery.FilterOption;
import com.epson.epos2.printer.Printer;
import com.epson.epos2.printer.PrinterStatusInfo;
import com.epson.epos2.printer.ReceiveListener;

public class epos2Plugin extends CordovaPlugin {
    private static final String TAG = "epos2";
    private CallbackContext callbackContext = null;
    private Printer printer = null;

    public void initialize(CordovaInterface cordova, CordovaWebView webView) {
        super.initialize(cordova, webView);
    }

    public boolean execute(final String action, final JSONArray args, final CallbackContext callbackContext) throws JSONException {
        this.callbackContext = callbackContext;

        if (action.equals("startDiscover")) {
            startDiscovery(callbackContext);
        } else if (action.equals("stopDiscover")) {
            stopDiscovery(callbackContext);
        } else if (action.equals("connectPrinter")) {
            connectPrinter(args, callbackContext);
        } else if (action.equals("disconnectPrinter")) {
            disconnectPrinter();
        } else if (action.equals("print")) {
            print(args, callbackContext);
        }

        return true;
    }

    private void startDiscovery(final CallbackContext callbackContext) {
        FilterOption mFilterOption = new FilterOption();
        mFilterOption.setDeviceType(Discovery.TYPE_PRINTER);
        mFilterOption.setEpsonFilter(Discovery.FILTER_NAME);
        try {
            Discovery.start(webView.getContext(), mFilterOption, discoveryListener);
        } catch (Epos2Exception e) {
            Log.e(TAG, "Error discovering printer: " + e.getErrorStatus(), e);
            callbackContext.error("Error discovering printer: " + e.getErrorStatus());
        }
    }

    private void stopDiscovery(final CallbackContext callbackContext) {
        while (true) {
            try {
                Discovery.stop();
                PluginResult result = new PluginResult(Status.OK, true);
                callbackContext.sendPluginResult(result);
                break;
            }
            catch (Epos2Exception e) {
                if (e.getErrorStatus() != Epos2Exception.ERR_PROCESSING) {
                    PluginResult result = new PluginResult(Status.ERROR, false);
                    callbackContext.sendPluginResult(result);
                    return;
                }
            }
        }
    }

    private void connectPrinter(final JSONArray args, final CallbackContext callbackContext) {
        Log.d(TAG, "Connecting Printer");

        try {
            printer = new Printer(Printer.TM_U220, Printer.MODEL_ANK, webView.getContext());
            printer.setReceiveEventListener(receiveListener);
        }
        catch (Epos2Exception e) {
            callbackContext.error("Error creating printer: " + e.getErrorStatus());
            Log.e(TAG, "Error creating printer: " + e.getErrorStatus(), e);
            return;
        }

        try {
            String target = args.getString(0);
            printer.connect("TCP:" + target , Printer.PARAM_DEFAULT);
        } catch (Epos2Exception e) {
            callbackContext.error("Error connecting printer: " + e.getErrorStatus());
            Log.e(TAG, "Error connecting printer: " + e.getErrorStatus(), e);
            return;
        } catch (JSONException e) {
            callbackContext.error("Error getting target: " + e.getCause());
            Log.e(TAG, "Error connecting printer", e);
            return;
        }

        try {
            printer.beginTransaction();
        } catch (Epos2Exception e) {
            callbackContext.error("Error beginning transaction");
            Log.e(TAG, "Error beginning transaction", e);
            return;
        }

        PluginResult result = new PluginResult(Status.OK, "Done connecting");
        callbackContext.sendPluginResult(result);
    }

    private void disconnectPrinter() {
        if (printer == null) {
            return;
        }

        try {
            printer.endTransaction();
        }
        catch (Epos2Exception e) {
            callbackContext.error("Error ending transaction: " + e.getErrorStatus());
            Log.e(TAG, "Error ending transaction: " + e.getErrorStatus(), e);
            return;
        }

        try {
            printer.disconnect();
        }
        catch (Epos2Exception e) {
            callbackContext.error("Error disconnecting printer: " + e.getErrorStatus());
            Log.e(TAG, "Error disconnecting printer: " + e.getErrorStatus(), e);
            return;
        }

        printer.clearCommandBuffer();
        printer.setReceiveEventListener(null);
        printer = null;

        PluginResult result = new PluginResult(Status.OK, true);
        callbackContext.sendPluginResult(result);
    }

    private void print(final JSONArray array, final CallbackContext callbackContext) {
        if (printer == null) {
            callbackContext.error("printer not found");
            return;
        }

        try {
            for (int i = 0; i < array.length(); i++) {
                String data = array.getString(i);
                if ("\n".equals(data)) {
                    printer.addFeedLine(1);
                } else {
                    printer.addText(data);
                }
            }

            printer.addCut(Printer.CUT_FEED);

            PrinterStatusInfo status = printer.getStatus();

            if (!isPrintable(status)) {
                callbackContext.error("Error printing: printer is not printable");
                Log.e(TAG, "Error printing: printer is not printable");

                try {
                    printer.disconnect();
                }
                catch (Epos2Exception ex) {
                    callbackContext.error("Error disconnecting");
                    Log.e(TAG, "Error disconnecting", ex);
                }
                return;
            }

            printer.sendData(Printer.PARAM_DEFAULT);
        } catch (Epos2Exception e) {
            callbackContext.error("Error printing");
            Log.e(TAG, "Error printing", e);
            try {
                printer.disconnect();
            }
            catch (Epos2Exception ex) {
                callbackContext.error("Error disconnecting");
                Log.e(TAG, "Error disconnecting", ex);
            }
        } catch (JSONException e) {
            callbackContext.error("Error getting data");
            Log.e(TAG, "Error getting data", e);
        }
    }

    private DiscoveryListener discoveryListener = new DiscoveryListener() {
        @Override
        public void onDiscovery(final DeviceInfo deviceInfo) {
            JSONObject item = new JSONObject();

            try {
                item.put("deviceName", deviceInfo.getDeviceName());
                item.put("target", deviceInfo.getTarget());
                item.put("ipAddress", deviceInfo.getIpAddress());
                item.put("macAddress", deviceInfo.getMacAddress());
                item.put("deviceType", deviceInfo.getDeviceType());
                item.put("bdAddress", deviceInfo.getBdAddress());
            } catch (JSONException e) {
                callbackContext.error("Error building device info");
            }

            callbackContext.success(item);
        }
    };

    private ReceiveListener receiveListener = new ReceiveListener() {
        @Override
        public void onPtrReceive(final Printer printer, final int code, final PrinterStatusInfo status, final String printJobId) {
            Log.d(TAG, "print success. status: " + status.getErrorStatus());

            new Thread(new Runnable() {
                @Override
                public void run() {
                    disconnectPrinter();
                }
            }).start();
        }
    };

    private boolean isPrintable(PrinterStatusInfo status) {
        if (status == null) {
            return false;
        }

        if (status.getConnection() == Printer.FALSE) {
            return false;
        }
        else if (status.getOnline() == Printer.FALSE) {
            return false;
        }

        return true;
    }
}
